Feat: Add logic and test code for `_GraphRolesMixin` as a parent class of `ADMG`

daehyun99 opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [X] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [X] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [X] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Ref #2280 (partially fixes it)

### List of changes to the codebase in this pull request
- Test: Verify `_GraphRolesMixin` as a parent class of `ADMG` 
- Feat: Add `_GraphRolesMixin` as a parent class of `ADMG`

daehyun99 committed:
Test: Verify `_GraphRolesMixin` as a parent class of `ADMG` (#2280)

- Modified the `test_empty_initialization` method in the `TestADMGInitialization` class.
- Created the `test_initialization_with_roles` method in the `TestADMGInitialization` class.
  - Test initialization with roles variables
- Modified the `test_is_m_separated` method in the `TestADMGSeparation` class.
- remove `import networkx as nx`

daehyun99 committed:
Feat: Add `_GraphRolesMixin` as a parent class of `ADMG` (#2280)

- Added code to initialize `roles` in `__init__()`.
- Enabled the use of the `roles` variable by adding `**attr` to `add_node(self, node, **attr)`.
- Removed unused imports: `from typing import Hashable, Optional, Sequence`.

ankurankan commented on pgmpy/base/ADMG.py:
```diff
@@ -35,11 +48,20 @@ def __init__(self, directed_ebunch=None, bidirected_ebunch=None, latents=None):
+        for role, vars in roles.items():
+            self.with_role(role=role, variables=vars, inplace=True)
+
+    def add_node(self, node, **attr):
```
This change is not required. Correct?

daehyun99 replied:
@ankurankan 
The following code snippet (lines 57-58) is essential for adding `roles` values to the `ADMG` class.

```python
        for role, vars in roles.items():
            self.with_role(role=role, variables=vars, inplace=True)
```

This implementation pattern is consistent across the `DAG` and `PDAG` classes when they also handle `roles` values. From a consistency perspective, it seems best to maintain this approach.

The `self.with_role` method internally uses `add_node`.

```python
def add_node(self, node, **attr):
```

As I understand it, This `add_node` method is originally defined in the `networkx.DiGraph` class, which the `ADMG`, `DAG`, and `PDAG` classes all inherit from.

- [`DiGraph`'s `add_node` method](https://github.com/networkx/networkx/blob/36e8a1ee85ca0ab4195a486451ca7d72153e2e00/networkx/classes/digraph.py#L439)

The problem is that the `DiGraph` class's `add_node` method does not have `roles` defined as a parameter.

Because of this, the `pgmpy` library's `DAG` class's `add_node` method handles `roles` by receiving them as `**kwargs`.

- [`DAG`'s `add_node` method](https://github.com/pgmpy/pgmpy/blob/d25b841e0e92f666a75c1f27d5341f4e3d680321/pgmpy/base/DAG.py#L284)

In contrast, the `PDAG` class doesn't define its own `add_node` method and instead directly inherits the one from `DiGraph`.

This leaves us with two potential options for the `ADMG` class:

1.  **Delete the `add_node` method from the `ADMG` class:** This would make its implementation consistent with the `PDAG` class.
2.  **Add a `**attr` argument to the `ADMG` class's `add_node` method:** This would align it with the `DAG` class's implementation.

My suggestion is to first ask the developer who added the `add_node` method to the `ADMG` class for their reasoning. Understanding their original intent will help us make the best decision. @Spinachboul

Thanks.

Spinachboul replied:
@daehyun99  I feel the `add_node` function which is there is `ADMG` class can still stretch itself to handle different attributes such as `roles` which would make it fully compatible with the `DAG` class.

daehyun99 replied:
@ankurankan Once you reply to this, I will proceed as requested.

ankurankan replied:
@daehyun99 @Spinachboul I had missed that the method signature for `ADMG` had deviated from `DAG`. I think we can keep the `attr` but rename it to `kwargs` to make it consistent with rest of the package. I would also suggest adding the `latent` argument ot both `ADMG.add_node` and `ADMG.add_nodes_from` as well similar to `DAG` methods. @daehyun99 Could you please also add the `latent` argument here?

daehyun99 replied:
@ankurankan I understand.

ankurankan commented on pgmpy/base/ADMG.py:
```diff
@@ -35,11 +48,20 @@ def __init__(self, directed_ebunch=None, bidirected_ebunch=None, latents=None):
         Adds a node to the ADMG from the MultiDiGraph class.
         """
-        super().add_node(node)
+        super().add_node(node, **attr)
```
Same here.

daehyun99 replied:
The content is included in the reply above.

ankurankan commented on pgmpy/tests/test_base/test_ADMG.py:
```diff
@@ -265,6 +276,8 @@ def test_is_m_separated(self):
 
         # Test with conditional set
         separated = self.admg.is_mseparated("A", "D", conditional_set={"C"})
+        assert separated is True
```
We can simplify this by not assigning the new variable and directly calling the assert on the expression. Similar to line 280

daehyun99 replied:
I've committed the updates to this part.
```py
        # Test with conditional set
        assert self.admg.is_mseparated("A", "D", conditional_set={"C"}) is True
        assert self.admg.is_mseparated("A", "D", conditional_set=set()) is False
```

ankurankan reviewed (changes requested):
@daehyun99 Thanks for the PR. I have added a couple of minor comments, but looks good to me otherwise.

daehyun99 committed:
Refactor: Simplify `test_is_m_separated` test assert (#2280)

- Apply code review feedback for PR #2294

daehyun99 committed:
Refactor: Changed `**attr` parameter to `**kwargs`. (#2280)

- Apply code review feedback for PR #2294

daehyun99 committed:
Feat: Add a `latent` parameter to the `add_node` and `add_nodes_from` methods of the `ADMG` class. (#2280)

- Apply code review feedback for PR #2294
- `add_node` method
	- Feat: Add the `latent` parameter.
	- Feat: Define the type for the `node` parameter.
	- Docs: Add comments for the parameters.
	- Refactor: Match the code style with the `add_node` method of the `DAG` class.
- `add_nodes_from` method
	- Feat: Add the `latent` parameter.
	- Feat: Define the type for the `node` parameter.
	- Docs: Add comments for the parameters.
	- Refactor: Match the code style with the `add_nodes_from` method of the `DAG` class.

ankurankan reviewed (approved):
Looks good to me. Thanks!!
